From 20ac03452be45af8d0b7275eee7883cf69fb12a1 Mon Sep 17 00:00:00 2001
From: William Douglas <william.douglas@intel.com>
Date: Mon, 25 Jul 2022 11:55:30 -0700
Subject: [PATCH] Force trapping math

---
 numpy/distutils/ccompiler.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/numpy/distutils/ccompiler.py b/numpy/distutils/ccompiler.py
index 8697fae..e11dd89 100644
--- a/numpy/distutils/ccompiler.py
+++ b/numpy/distutils/ccompiler.py
@@ -391,7 +391,7 @@ def CCompiler_customize_cmd(self, cmd, ignore=()):
     log.info('customize %s using %s' % (self.__class__.__name__,
                                         cmd.__class__.__name__))
 
-    if hasattr(self, 'compiler') and 'clang' in self.compiler[0]:
+    if hasattr(self, 'compiler'):
         # clang defaults to a non-strict floating error point model.
         # Since NumPy and most Python libs give warnings for these, override:
         self.compiler.append('-ftrapping-math')
-- 
2.36.1

