diff --git a/numpy/distutils/unixccompiler.py b/numpy/distutils/unixccompiler.py
index fb91f17..e362e74 100644
--- a/numpy/distutils/unixccompiler.py
+++ b/numpy/distutils/unixccompiler.py
@@ -52,6 +52,11 @@ def UnixCCompiler__compile(self, obj, src, ext, cc_args, extra_postargs, pp_opts
     try:
         self.spawn(self.compiler_so + cc_args + [src, '-o', obj] + deps +
                    extra_postargs, display = display)
+        self.spawn(self.compiler_so + cc_args + ["-march=haswell", "-O3", "-fno-semantic-interposition", "-ffat-lto-objects", "-flto=auto"] + [src, '-o', obj + ".avx2"] + deps +
+                   extra_postargs, display = display)
+        self.spawn(self.compiler_so + cc_args + ["-march=skylake-avx512", "-O3", "-fno-semantic-interposition", "-ffat-lto-objects", "-flto=auto", "-mprefer-vector-width=256"] + [src, '-o', obj + ".avx512"] + deps +
+                   extra_postargs, display = display)
+
     except DistutilsExecError as e:
         msg = str(e)
         raise CompileError(msg) from None
